File: backend/test_observer/controllers/issues/attachment_rules.py
# Copyright (C) 2023 Canonical Ltd.
#
# This file is part of Test Observer Backend.
#
# Test Observer Backend is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License version 3, as
# published by the Free Software Foundation.
#
# Test Observer Backend is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


from fastapi import APIRouter, Depends, HTTPException, Security
from sqlalchemy.orm import Session

from test_observer.common.permissions import Permission, permission_checker
from test_observer.data_access.setup import get_db
from test_observer.data_access.models import (
    IssueTestResultAttachmentRule,
    Issue,
    IssueTestResultAttachmentRuleExecutionMetadata,
)

from .models import (
    IssueTestResultAttachmentRulePostRequest,
    IssueTestResultAttachmentRulePatchRequest,
)

from .shared_models import MinimalIssueTestResultAttachmentRuleResponse

router = APIRouter()


@router.post(
    "/{issue_id}/attachment-rules",
    response_model=MinimalIssueTestResultAttachmentRuleResponse,
    dependencies=[
        Security(permission_checker, scopes=[Permission.change_attachment_rule])
    ],
)
def post_attachment_rule(
    issue_id: int,
    request: IssueTestResultAttachmentRulePostRequest,
    db: Session = Depends(get_db),
):
    # Get the issue
    issue = db.get(Issue, issue_id)
    if issue is None:
        raise HTTPException(status_code=404, detail="Issue not found")

    # Create the attachment rule
    attachment_rule = IssueTestResultAttachmentRule(
        issue_id=issue.id,
        enabled=request.enabled,
        families=request.families,
        environment_names=request.environment_names,
        test_case_names=request.test_case_names,
        template_ids=request.template_ids,
        execution_metadata=[
            IssueTestResultAttachmentRuleExecutionMetadata(
                category=metadata.category,
                value=metadata.value,
            )
            for metadata in request.execution_metadata.to_rows()
        ],
    )

    # Add the attachment rule
    db.add(attachment_rule)
    db.commit()
    db.refresh(attachment_rule)
    return attachment_rule


@router.patch(
    "/{issue_id}/attachment-rules/{attachment_rule_id}",
    response_model=MinimalIssueTestResultAttachmentRuleResponse,
    dependencies=[
        Security(permission_checker, scopes=[Permission.change_attachment_rule])
    ],
)
def patch_attachment_rule(
    issue_id: int,
    attachment_rule_id: int,
    request: IssueTestResultAttachmentRulePatchRequest,
    db: Session = Depends(get_db),
):
    # Get the attachment rule
    attachment_rule = db.get(IssueTestResultAttachmentRule, attachment_rule_id)
    if attachment_rule is None:
        raise HTTPException(status_code=404, detail="Attachment rule not found")
    if attachment_rule.issue_id != issue_id:
        raise HTTPException(
            status_code=400, detail="Attachment rule not attached to given issue"
        )

    # Modify the attachment rule
    if request.enabled is not None:
        attachment_rule.enabled = request.enabled

    # Save the attachment rule
    db.commit()
    db.refresh(attachment_rule)
    return attachment_rule


@router.delete(
    "/{issue_id}/attachment-rules/{attachment_rule_id}",
    status_code=204,
    dependencies=[
        Security(permission_checker, scopes=[Permission.change_attachment_rule])
    ],
)
def delete_attachment_rule(
    issue_id: int,
    attachment_rule_id: int,
    db: Session = Depends(get_db),
):
    # Get the attachment rule
    attachment_rule = db.get(IssueTestResultAttachmentRule, attachment_rule_id)
    if attachment_rule is None:
        return
    if attachment_rule.issue_id != issue_id:
        raise HTTPException(
            status_code=400, detail="Attachment rule not attached to given issue"
        )

    # Delete the attachment rule
    db.delete(attachment_rule)
    db.commit()

File: frontend/lib/ui/issue_page/attachment_rules.dart
// Copyright 2025 Canonical Ltd.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License version 3, as
// published by the Free Software Foundation.
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// SPDX-FileCopyrightText: Copyright 2025 Canonical Ltd.
// SPDX-License-Identifier: GPL-3.0-only

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:dartx/dartx.dart';

import 'blinking_content.dart';

import '../../models/attachment_rule.dart';
import '../../providers/issue.dart';
import '../../routing.dart';
import '../attachment_rule.dart';
import '../expandable.dart';

class AttachmentRulesSection extends StatelessWidget {
  const AttachmentRulesSection({
    super.key,
    required this.issueId,
    required this.attachmentRules,
    this.expandRuleId,
  });

  final int issueId;
  final List<AttachmentRule> attachmentRules;
  final int? expandRuleId;

  @override
  Widget build(BuildContext context) {
    // Retrieve initially expanded attachment rule from URL
    final uri = AppRoutes.uriFromContext(context);
    final int? focusOnRuleId =
        uri.queryParameters['attachmentRule']?.toIntOrNull();

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          'Attachment Rules',
          style: Theme.of(context).textTheme.headlineMedium,
        ),
        ...attachmentRules.map(
          (rule) => AttachmentRuleExpandable(
            issueId: issueId,
            attachmentRule: rule,
            focusOnRule: focusOnRuleId != null && rule.id == focusOnRuleId,
          ),
        ),
      ],
    );
  }
}

class AttachmentRuleExpandable extends ConsumerWidget {
  const AttachmentRuleExpandable({
    super.key,
    required this.issueId,
    required this.attachmentRule,
    this.focusOnRule = false,
  });

  final int issueId;
  final AttachmentRule attachmentRule;
  final bool focusOnRule;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return BlinkingContent(
      blink: focusOnRule,
      child: Expandable(
        initiallyExpanded: focusOnRule,
        title: Row(
          children: [
            Text('Attachment Rule #${attachmentRule.id}'),
            const Spacer(),
            TextButton(
              child: Text(attachmentRule.enabled ? 'disable' : 'enable'),
              onPressed: () async {
                if (attachmentRule.enabled) {
                  await ref
                      .read(issueProvider(issueId).notifier)
                      .disableAttachmentRule(
                        issueId: issueId,
                        attachmentRuleId: attachmentRule.id,
                      );
                } else {
                  await ref
                      .read(issueProvider(issueId).notifier)
                      .enableAttachmentRule(
                        issueId: issueId,
                        attachmentRuleId: attachmentRule.id,
                      );
                }
              },
            ),
            TextButton(
              child: Text('delete'),
              onPressed: () {
                showDialog<void>(
                  context: context,
                  builder: (context) => AlertDialog(
                    title: const Text('Delete Attachment Rule'),
                    content: const Text(
                      'Are you sure you want to delete this attachment rule?',
                    ),
                    actions: [
                      TextButton(
                        onPressed: () => Navigator.of(context).pop(),
                        child: const Text('Cancel'),
                      ),
                      TextButton(
                        onPressed: () async {
                          await ref
                              .read(issueProvider(issueId).notifier)
                              .deleteAttachmentRule(
                                issueId: issueId,
                                attachmentRuleId: attachmentRule.id,
                              );
                          if (context.mounted) {
                            Navigator.of(context).pop();
                          }
                        },
                        child: const Text('Delete'),
                      ),
                    ],
                  ),
                );
              },
            ),
          ],
        ),
        children: [
          Padding(
            padding: const EdgeInsets.symmetric(horizontal: 24.0),
            child: AttachmentRuleFiltersWidget(
              filters: attachmentRule.toFilters(),
              editable: false,
            ),
          ),
        ],
      ),
    );
  }
}
